pika/adapters/select_connection.py

```diff
@@ -264,21 +264,38 @@ def get_next_deadline(self):
         timeout = min((timeout, SelectPoller.POLL_TIMEOUT))
         return timeout * SelectPoller.POLL_TIMEOUT_MULT
 
+    @staticmethod
+    def is_ready_to_fire(to_run):
+        """For patching by unittest"""
+        return True
+
     def process_timeouts(self):
         """Process the self._timeouts event stack"""
 
         now = time.time()
-        to_run = [timer for timer in self._timeouts.values()
-                  if timer['deadline'] <= now]
-
         # Run the timeouts in order of deadlines. Although this shouldn't
         # be strictly necessary it preserves old behaviour when timeouts
         # were only run periodically.
-        for t in sorted(to_run, key=itemgetter('deadline')):
-            t['callback']()
-            del self._timeouts[hash(frozenset(t.items()))]
+        to_run = sorted([(k,timer) for (k,timer) in self._timeouts.items()
+                         if timer['deadline'] <= now],
+                        key = lambda item : item[1]['deadline'])
+
+        if self.is_ready_to_fire(to_run):
```

Comment:
Remove the `if self.is_ready_to_fire(to_run):` check. This pollutes the production code, and it's not necessary for making the unit tests work. See my comment on `test_timers_for_murder` that describes how to implement the test without this.